rules/no-useless-promise-resolve-reject.js

```diff
@@ -0,0 +1,120 @@
+	[RETURN_REJECT]: 'Prefer `throw error` over `return Promise.reject(error)`.',
+	[YIELD_RESOLVE]: 'Prefer `yield value` over `yield Promise.resolve(value)`.',
+	[YIELD_REJECT]: 'Prefer `throw value` over `yield Promise.reject(value)`.',
+};
+
+const promiseResolveOrRejectSelector = methodCallSelector({
+	object: 'Promise',
+	methods: ['resolve', 'reject'],
+	argumentsLength: 1,
+});
+const asyncArrowFunctionReturnSelector = `ArrowFunctionExpression[async=true] > ${promiseResolveOrRejectSelector}.body`;
+const returnStatementSelector = `ReturnStatement > ${promiseResolveOrRejectSelector}.argument`;
+const yieldExpressionSelector = `YieldExpression > ${promiseResolveOrRejectSelector}.argument`;
+
+const functionTypes = new Set([
+	'ArrowFunctionExpression',
+	'FunctionDeclaration',
+	'FunctionExpression',
+]);
+const getParentFunction = node => {
+	let {parent} = node;
+	while (parent && !functionTypes.has(parent.type)) {
+		parent = parent.parent;
+	}
+
+	return parent;
+};
+
+/** @param {import('eslint').Rule.RuleContext} context */
+const create = context => {
+	const sourceCode = context.getSourceCode();
+
+	/** @param {(isResolve: boolean, valueString: string, value) => import('eslint').Rule.ReportFixer} createFix */
+	const createProblem = (node, createFix, resolveMessage = RETURN_RESOLVE, rejectMessage = RETURN_REJECT) => {
+		const isResolve = node.callee.property.name === 'resolve';
+		const value = node.arguments.length === 1 ? node.arguments[0] : undefined;
+		return {
+			node: node.callee,
+			messageId: isResolve ? resolveMessage : rejectMessage,
+			fix: value === undefined ? undefined : createFix(isResolve, sourceCode.getText(value), value),
+		};
+	};
+
+	return {
+		[asyncArrowFunctionReturnSelector](node) {
+			return createProblem(
+				node,
+				(isResolve, valueString, value) => fixer =>
+					fixer.replaceText(
+						node,
+						isResolve
+							// Turns `=> Promise.resolve(value)` into `=> value`
+							? (value.type === 'ObjectExpression' || value.type === 'SequenceExpression'
+								? `(${valueString})`
+								: valueString
+							)
+							// Turns `=> value` into `=> { throw value }`
+							: `{ throw ${valueString}; }`,
+					),
+			);
+		},
+		[returnStatementSelector](node) {
+			const parentFunction = getParentFunction(node);
+			if (!parentFunction || !parentFunction.async) {
+				return;
+			}
+
+			return createProblem(
+				node,
+				(isResolve, valueString) => fixer => isResolve
+					// Turns `return Promise.resolve(value)` into `return value`
+					? fixer.replaceText(node, valueString)
+					// Turns `return Promise.reject(value)` into `throw value`
+					: fixer.replaceText(node.parent, `throw ${valueString};`),
+			);
+		},
+		[yieldExpressionSelector](node) {
+			const parentFunction = getParentFunction(node);
+			if (!parentFunction || !parentFunction.async || !parentFunction.generator) {
+				return;
+			}
+
+			return createProblem(
```

Comment:
We need check `YieldExpression.delegate` here, we can't fix

```js
yield* Promise.resolve([])
```

Reply:
We also need check if `YieldExpression` is an `ExpressionStatement`, because `YieldExpression` can be any where

```js
async function * a() {
  alert(yield Promise.resolve(1));
}
```

```js
async function * a() {
  const foo = yield Promise.resolve(1);
}
```

Reply:
Why do we need to check if it's in an `ExpressionStatement`? `f` and `g` below seem to behave the same:

```js
async function* f() {
  console.log('next:', yield Promise.resolve(1))
}

async function* g() {
  console.log('next:', yield 1)
}

const a = f()
const a1 = await a.next()
console.log(a1)            // {value: 1, done: false}
const a2 = await a.next(2) // next: 2
console.log(a2)            // {value: undefined, done: true}

const b = g()
const b1 = await b.next()
console.log(b1)            // {value: 1, done: false}
const b2 = await b.next(2) // next: 2
console.log(b2)            // {value: undefined, done: true}
```

Reply:
Sorry, I mean `Promise.reject()`